### PR TITLE
Enable GridOut::write_gmsh() for simplex meshes

### PR DESCRIPTION
**Pull request by bangerth**
I decided to put some of the necessary infrastructure into `ReferenceCell`.

/rebuild

### CODE REVIEW

File: source/grid/grid_out.cc [hunk 1]
```diff
@@ -3951,29 +3912,25 @@ GridOut::write_msh_faces(const Triangulation<dim, spacedim> &tria,
...
+            if (face->reference_cell() == ReferenceCells::Quadrilateral)
+              out << ' '
+                  << face->vertex_index(
+                       GeometryInfo<dim - 1>::ucd_to_deal[vertex]) +
+                       1;
+            else if ((face->reference_cell() == ReferenceCells::Triangle) ||
+                     (face->reference_cell() == ReferenceCells::Line))
+              out << ' ' << face->vertex_index(vertex) + 1;
+            else
+              Assert(false, ExcInternalError());
```
Comment (drwells) [inline source/grid/grid_out.cc:3932]: If we are converting the other places, could we do `ReferenceCell::gmsh_vertex_to_deal_vertex()` too?

File: tests/simplex/grid_out_msh.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    // For the cylinder geometry
+    for (auto cell : x.active_cell_iterators())
+      for (const unsigned int face : cell->face_indices())
+        if (cell->at_boundary(face))
+          if (std::fabs(cell->face(face)->center()[0] - 0) <
+              0.1 * cell->diameter())
+            cell->face(face)->set_boundary_id(1);
+          else if (std::fabs(cell->face(face)->center()[0] - 4) <
+                   0.1 * cell->diameter())
+            cell->face(face)->set_boundary_id(2);
```
Comment (drwells) [inline tests/simplex/grid_out_msh.cc:46]: I get an ambiguous brace warning here:
```suggestion
          {
            if (std::fabs(cell->face(face)->center()[0] - 0) <
                0.1 * cell->diameter())
              cell->face(face)->set_boundary_id(1);
            else if (std::fabs(cell->face(face)->center()[0] - 4) <
                     0.1 * cell->diameter())
              cell->face(face)->set_boundary_id(2);
           }
```

**Review by drwells — CHANGES_REQUESTED** [review]

**Comment by drwells** [comment]
Its not directly related to this PR (the same problems occurs with VTK) but I see the following when I plot the output:

weird-cylinder

I think this means that point placement currently ignores manifold descriptions.

File: source/grid/grid_out.cc (hunk 1 above)
Comment (bangerth) [reply source/grid/grid_out.cc:3932]: I could do that. Would you allow me to leave this for a follow-up?

**Comment by bangerth** [comment]
I fixed the one place the tester complains about.

As for the figure: This looks correct to me. Remember that I'm outputting the mesh with linear tetrahedra. So only the vertex positions are right, but lines are shown as straight whether they are or are not based on some (deal.II-internal) manifold description. The vertex positions look correct to me: all on the perimeter of the cylinder.

File: source/grid/grid_out.cc (hunk 1 above)
Comment (drwells) [reply source/grid/grid_out.cc:3932]: sure.

**Comment by drwells** [comment]
Yes, I think you're right. I'll double-check the grid later.

**Review by drwells — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Pull request merged**